### ISSUE TITLE
FI_FENCE: desire for stronger semantics

### ISSUE THREAD

**Issue by shefty** [ofiwg/libfabric#3186]
FI_FENCE is limited to operations that target the same peer EP.  For unconnected EPs, this limits its usefulness.  A full barrier semantic seems to be preferable.

Note that we can define FI_FENCE with stronger semantics without breaking backwards compatibility, if we don't want to introduce a new flag.

**Comment by jswaro** [comment]
I'd be concerned with changing this behavior to be a fabric-wide fence. 

This would negatively affect message transmission rates when an application may only need a EP specific fence. 

Perhaps a second fence, a global fabric-level fence would be desirable.

**Comment by shefty** [comment]
For reference, these are the requirements for MPI and SHMEM.  Other apps may have other requirements for fencing, especially those using connections.
---
MPI_Win_fence (collective call this one is slighty different from the rest): The MPI call MPI_WIN_FENCE(assert, win) synchronizes RMA calls on win. The call is collective on the group of win. All RMA operations on win originating at a given process and started before the fence call will complete at that process before the fence call returns. They will be completed at their target before the fence call returns at the target. RMA operations on win started by a process after the fence call returns will access their target window only after MPI_WIN_FENCE has been called by the target process.

MPI_Win_flush: MPI_WIN_FLUSH completes all outstanding RMA operations initiated by the calling process to the target rank on the specified window. The operations are completed both at the origin and at the target.

MPI_Win_flush_all: All RMA operations issued by the calling process to any target on the specified window 17 prior to this call and in the specified window will have completed both at the origin and at 18 the target when this call returns.

MPI_Win_start/MPI_Win_complete: 
       1. MPI_Win_start: Starts an RMA access epoch for win. RMA calls issued on win during this epoch must access only windows at processes in group. Each process in group must issue a matching call to MPI_WIN_POST. RMA accesses to each target window will be delayed, if necessary, until the target process executed the matching call to MPI_WIN_POST. MPI_WIN_START is allowed to block until the corresponding MPI_WIN_POST calls are executed, but is not required to.
	2. MPI_Win_complete: Completes an RMA access epoch on win started by a call to MPI_WIN_START. All RMA communication calls issued on win during this epoch will have completed at the origin when the call returns.

MPI_Win_lock/MPI_Win_unlock:
	1. MPI_Win_lock: Starts an RMA access epoch. The window at the process with rank rank can be accessed by RMA operations on win during that epoch. Multiple RMA access epochs (with calls to MPI_WIN_LOCK) can occur simultaneously; however, each access epoch must target a different process.
	2. Completes an RMA access epoch started by a call to MPI_WIN_LOCK on window win. RMA operations issued during this period will have completed both at the origin and at the target when the call returns.

SHMEM fence: This routine assures ordering of delivery of Put, AMOs, and memory store routines to symmetric data objects. All Put, AMOs, and memory store routines to symmetric data objects issued to a particular remote PE prior to the call to shmem_fence are guaranteed to be delivered before any subsequent Put, AMOs, and memory store routines to symmetric data objects to the same PE. shmem_fence guarantees order of delivery, not completion.

**Comment by shefty** [comment]
This is the current FI_FENCE definition, which is definitely vague.
```
FI_FENCE
Applies to transmits. Indicates that the requested operation, also known as the fenced operation, be deferred until all previous operations targeting the same target endpoint have completed.
```
I *think* the issue is that the behavior of the operation posted immediately after the fenced operation is unclear.  The current definition suggests that the un-fenced operation may pass the fenced operation.  The desired behavior is that the fence acts as a barrier at least to the targeted peer.  That is, no operations posted after the fenced operation may proceed until prior operations complete.  The ordering of operations posted after fencing is determined using whatever ordering semantics are defined for the endpoint.

A proposed semantic is:
```
FI_SUPER_FENCE
Applies to transmits.  Indicates that the requested operation, also known as the fenced operation, and any operation posted after the fenced operation will be deferred until all previous operations targeting the same peer endpoint have completed.  Operations posted after the fencing will see and/or replace the results of any operations initiated prior to the fenced operation.

The ordering of operations starting at the posting of the fenced operation (inclusive) to the posting of a subsequent fenced operation (exclusive) is controlled by the endpoint's ordering semantics.
```
If this looks reasonable, then we can determine if it's acceptable to strengthen the current FI_FENCE semantic, or if a new transmit flag is needed.  It's possible that applications are actually assuming the FI_SUPER_FENCE behavior.  Based on a quick look, I think this is what the providers are implementing.

**Comment by sayantansur** [comment]
+1 on the FI_SUPER_FENCE! I think it fits SHMEM well. I'll have Jim @jdinan @jeffhammond review.

**Comment by jswaro** [comment]
> I think the issue is that the behavior of the operation posted immediately after the fenced operation is unclear.

I would agree that it is unclear whether the fenced operation itself could be passed by operations posted after it. This should be clarified so that applications can be clear on expected behavior. 

> If this looks reasonable, then we can determine if it's acceptable to strengthen the current FI_FENCE semantic, or if a new transmit flag is needed.

I'll admit to being slightly confused. I thought originally that application developers were requesting a broader fence? This proposed definition of 'super fence' appears to be a more strongly worded version of FI_FENCE. Note: I am not objecting to a more strongly defined definition.

**Comment by shefty** [comment]
What I mentioned in the ofiwg wasn't what was needed.  That was my mistake.  The real issue seems to be if an operation posted after the fenced operation can pass the fenced call.  It sounds like the fence definition itself is okay (fenced to the same peer endpoint).

**Comment by jeffhammond** [comment]
We are going to build a fence. It's going to be a super fence. The greatest
fence you've ever seen.

But seriously, where are the details? I have no context for this proposal.
-- 
Jeff Hammond
jeff.science@gmail.com
http://jeffhammond.github.io/

**Comment by shefty** [comment]
The issue is that we're building a fence, but going to call it a wall because that sounds better.

The core problem is: ```the behavior of the operation posted immediately after the fenced operation is unclear. The current definition suggests that the un-fenced operation may pass the fenced operation.```  Applications want the fence to be a barrier between all ops posted before it and all ops posted after it, not just the fencing of a marked operation.

To make use of the current definition, an app would need to set the FI_FENCE flag on *all* operations in order to get the desired application semantics. That would result in a significant performance hit.

I prefer to 'fix' this by adding clarifying comments about FI_FENCE, rather than actually defining a new FI_SUPER_FENCE op flag.

**Comment by jdinan** [comment]
FI_SUPER_FENCE appears to address the basic semantic problem, which is that FI_FENCE can't provide the right level of ordering for subsequent operations.  There is a lingering challenge, which is that shmem_fence guarantees ordering between the caller and all peers (i.e. it takes no arguments).  To use FI_SUPER_FENCE, I think the middleware would need to keep track of which peers had been fenced and apply FI_SUPER_FENCE on the next operation targeting each peer (and we can obviously reset the fence-pending state when we see things like barriers and quiets).

Are there cases where we expect this to be more efficient than what we can currently build on top of libfabric using event counters and triggered operations (e.g. http://dl.acm.org/citation.cfm?id=2866380)?  Or, in cases where the middleware knows that the fabric is ordered and reliable (i.e. fence is a no-op)?

In terms of the mechanics of the API, SOS uses more than one transmit endpoint to select from different behaviors (not a semantic requirement, but it improves performance).  We would need to choose whether to keep these optimizations or use FI_SUPER_FENCE with a single EP (or accept the overhead of fencing multiple EPs).

MPI RMA should be able to use the FI_FENCE flag as it currently exists to enforce accumulate ordering.  It is helpful that you can choose which operations are ordered rather than getting an all-subsequent-operations semantic (e.g. put/get in the same epoch are not ordered relative to accumulates).  The synchronization operations you mentioned above all more or less require the implementation to guarantee remote completion, which is stronger than what FI_SUPER_FENCE provides.  Off the top of my head, I can't think of an MPI RMA use case for the new flag.

**Comment by sayantansur** [comment]
@jeffhammond in addition to what Sean said, my goal is to be able to have SHMEM use the FI_SUPER_FENCE semantic to implement shmem_fence(). Currently Sandia OpenSHMEM is waiting on all operations to -complete- before exiting shmem_fence(). While that is correct, it is over-synchronizing and losing performance.

https://github.com/Sandia-OpenSHMEM/SOS/blob/master/src/transport_ofi.h#L334

**Comment by jdinan** [comment]
@sayantansur We can do better with triggered ops and have published the algorithm.  IMHO this is an engineering problem, as much in libfabric (triggered ops support) as it is in SOS.

SOS also supports an ordered communication mode where fence is a no-op.  This is supported for Portals, but has bit rotted badly in the OFI build.  Making this work again is also an engineering problem.

I'm curious why you think this should be solved at the libfabric level?

**Comment by sayantansur** [comment]
Well, my reasons are:

1. Triggered operations are lower-level semantic and not a 1:1 map with the application semantic. If possible, libfabric should provide the application level semantic, and let the provider _do the right thing_.

2. Triggered operations might be easy for some providers to implement, not so easy for others, and other providers might lose performance.

I believe with FI_SUPER_FENCE the triggered operation based optimization can be moved inside the provider, thereby simplifying SHMEM, and other middleware that might need a similar semantic.

The question is whether FI_SUPER_FENCE is allowing us to do the right thing for SHMEM? Are there any modifications to the proposed definition that you'd like?

**Comment by jdinan** [comment]
Sure, determining how best to optimize shmem_fence above libfabric is not easy, since the middleware doesn't have any information about the relative performance of various libfabric API (e.g. if triggered ops are implemented in software).  As far as semantic match goes, see my previous message in this thread.  FI_SUPER_FENCE is almost, but not quite, a shmem_fence in a way that will create extra bookkeeping at the middleware layer.

**Comment by sayantansur** [comment]
@jdinan The SHMEM fence definition that is quoted above says that fence only applies per PE, even though the call is global in scope. Does that still require the bookkeeping in middleware?

**Comment by shefty** [comment]
I think there are 2 things needed:

1. FI_FENCE needs to provide stronger semantics (matching FI_SUPER_FENCE above).  The current semantics don't seem usable for anything.
2. We _may_ need a new FI_BARRIER (or FI_SUPER_DUPER_FENCE) flag that fences across all peer endpoints.

For 1, I'm also considering how fencing will work with connected endpoints and storage applications.

**Comment by jdinan** [comment]
@sayantansur - I believe so.  The middleware needs to enforce ordering for every peer it has targeted since the last ordering point.  Given that FI_SUPER_FENCE is point-to-point, the middleware would need to keep track of which PEs have been fenced to determine whether a given operation should have FI_SUPER_FENCE set.

@shefty - (1) Is FI_FENCE not strong enough for MPI RMA?  Seems to me like it should be.  What am I missing?

(2) I assume that shmem_quiet would still be done in the middleware layer (fi_cntr_wait).  A shmem_quiet implies shmem_fence, but the libfabric layer would have no way of knowing that a fence had been achieved.  Would you also have to replicate the event tracking that we are doing at the middleware layer?

**Comment by shefty** [comment]
Imagine this flow: read1, write2, write3.  With most rdma hardware, writes can pass reads.  In order to prevent this, the app needs to add FI_FENCE to write2 _and_ write3.  But that also fences write3 against write2, which may not be needed.  In fact, *every* write that occurs after read1 must be fenced with the current definition, or that write can jump ahead of read1.  I don't see how this is usable.

If FI_BARRIER is added, then one possible implementation would be to use an internal counter to track completions, similar to what an app would do.  However, a provider may have other ways of supporting this semantic based on the internal implementation.  It may only need to wait for the current transmit queue to drain before starting the next operation.  This could be done without the use of counters.  I'm not proposing adding FI_BARRIER, only mentioning it for discussion at the moment.

**Comment by sayantansur** [comment]
@jdinan I don't know how you can get around that even with the direct usage of triggered (which would mean you would have to use a counter per destination PE). One approach is to have *all* puts/gets triggered after the fence point. Which is probably the FI_SUPER_DUPER_FENCE that Sean brought up.

**Comment by jdinan** [comment]
@shefty You might be right that this ordering model is bad for most models, but for MPI RMA it actually sounds reasonable.  All MPI RMA operations are unordered (and one must assume can overtake) apart from accumulates (atomics).  The MPI runtime would apply FI_FENCE to the relevant accumulates and nothing else.  It gives you more ordering than you need, since accumulates may also get ordered relative to put/get operation in the same epoch, but it's still less ordering than you'd have with FI_SUPER_FENCE.  As an optimization, the MPI runtime could put accumulates and put/get on separate TX endpoints to eliminate the unnecessary ordering.

**Comment by jswaro** [comment]
Keep in mind, more broadly sweeping fences that affect all endpoints could have adverse effects of user applications in unexpected ways.

For example, consider the case when an application uses more than one middleware library, which both utilize libfabric. Since each library exists within the same process space, anything global to the process,such as a 'super duper fence', may affect the entire fabric unexpectedly. Suppose library one issues a super fence. Depending on implementation, it might cause transactions from library two to be halted unnecessarily until the super fence from library one completes. 

It may be safe to assume that applications might only have one libfabric consumer most of the time, but it's not a guarantee. 

The application has the right context to understand what endpoints need to be part of a broader fence, something a provider *may* not be privy to. Even if we would be privy to such information, by potentially just fencing all endpoints within a domain, the provider may end up essentially performing the same action(s) that an application might take anyway, but forcing the logic further down into providers.

**Comment by shefty** [comment]
So far there's no proposal to extend any fence semantic beyond a single local endpoint.  A fence operation on one endpoint should not affect traffic from another endpoint, unless the underlying implementation is goofy.

**Comment by jeffhammond** [comment]
Is it possible to combine multicast and fence? Could we then add a super fence that aggregates the behavior of fencing all the endpoints in a multicast group?

**Comment by shefty** [comment]
I'm not sure about using FI_MULTICAST specifically, but I like the idea that FI_FENCE could be combined with another flag to provide the stronger semantics.

For now, I'm still trying to figure out what changes need to be made to the existing FI_FENCE semantic.

**Comment by shefty** [comment]
See #3218 - updated FI_FENCE definition based on FI_SUPER_FENCE description above.

**Comment by jdinan** [comment]
This is definitely an improvement and may be good for MPI.  Because a shmem_fence requires ordering with respect to all peers (and FI_FENCE is point-to-point), I'm not sure I would use this method to implement it.  However, with the new semantics, at least I get the choice.  The problem is that using this method would put bookkeeping for which PEs have been communicated with on the critical path for a model that is message rate limited, with small payloads.

**Comment by sayantansur** [comment]
I also agree with Jim that this is an improvement, and gives us an option.

Jim - while you can avoid the book-keeping in your current global counter approach, we may be over-synchronizing in shmem_fence(). Example: PE 1 puts to PE 2 and PE 3. Then calls fence, followed by a Put to PE 2. Per the SHMEM semantics, as I understand, the second Put to PE 2 can proceed without waiting for the Put to PE 3 to finish. Is this a valid situation in the sense that apps might do?

**Comment by jdinan** [comment]
Correct, you get to choose between possibly over-synchronizing and possibly reducing your message rate.

**Comment by sayantansur** [comment]
OK, this is probably a SOS activity, we should probably look at how many instructions a bloom filter type technique would add, and what delta it might have on message rate.

**Comment by jdinan** [comment]
Inserting into the Bloom filter would be best case dozens of instructions depending on the number and complexity of the hash functions (which influence rate of false positives).  Assume that we dump the filter into a simple array of flags (NPES length) during the fence and reset the filter.  On each operation you also need to do a lookup in this flags array to check whether to set FI_FENCE on the operation.  So, total SWAG, I'd say no fewer than 50 instructions on a put.

**Comment by jeffhammond** [comment]
People who care about message rate should use unordered, nonblocking ops.
MPI-3 has those. SHMEM needs to get with the program 😁
-- 
Jeff Hammond
jeff.science@gmail.com
http://jeffhammond.github.io/

**Comment by sayantansur** [comment]
50 instructions would obviously be a non-starter. An array of flags per destination would be far cheaper in number of instructions, but might introduce cache misses in an app that does random small message communication. Another "hack" we could do is to bucket the PEs so we can limit the amount of over-synchronization. In any case, it seems FI_FENCE redefinition is good for now.

**Comment by jdinan** [comment]
@jeffhammond shmem_put_nbi??

@sayantansur Right, we can definitely get counts down.  We can also fall back to counter-based fence once we cross a number of peers threshold.  Any solution will also need to be viable in SHMEM_THREAD_MULTIPLE mode, which increases overheads.

**Comment by jeffhammond** [comment]
@jdinan I haven't paid much attention since OpenSHMEM 1.2 was released...
